controller/composite/metacontroller.go

```diff
@@ -17,153 +17,81 @@ limitations under the License.
 	mc := &Metacontroller{
+		k8s:          k8s,
 		resources:    resources,
 		mcClient:     mcClient,
 		dynClient:    dynClient,
 		dynInformers: dynInformers,
 
-		ccLister:         mcInformerFactory.Metacontroller().V1alpha1().CompositeControllers().Lister(),
-		ccInformer:       mcInformerFactory.Metacontroller().V1alpha1().CompositeControllers().Informer(),
 		revisionLister:   mcInformerFactory.Metacontroller().V1alpha1().ControllerRevisions().Lister(),
 		revisionInformer: mcInformerFactory.Metacontroller().V1alpha1().ControllerRevisions().Informer(),
 
-		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "CompositeController"),
 		parentControllers: make(map[string]*parentController),
 
 		numWorkers: numWorkers,
 	}
 
-	mc.ccInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc:    mc.enqueueCompositeController,
-		UpdateFunc: mc.updateCompositeController,
-		DeleteFunc: mc.enqueueCompositeController,
-	})
-
 	return mc
 }
 
-func (mc *Metacontroller) Start() {
-	mc.stopCh = make(chan struct{})
-	mc.doneCh = make(chan struct{})
-
-	go func() {
-		defer close(mc.doneCh)
-		defer utilruntime.HandleCrash()
-
-		klog.InfoS("Starting CompositeController metacontroller")
-		defer klog.InfoS("Shutting down CompositeController metacontroller")
-
-		if !cache.WaitForNamedCacheSync("CompositeController", mc.stopCh, mc.ccInformer.HasSynced) {
-			return
-		}
-
-		// In the metacontroller, we are only responsible for starting/stopping
-		// the actual controllers, so a single worker should be enough.
-		for mc.processNextWorkItem() {
-		}
-	}()
-}
-
-func (mc *Metacontroller) Stop() {
-	// Stop metacontroller first so there's no more changes to controllers.
-	close(mc.stopCh)
-	mc.queue.ShutDown()
-	<-mc.doneCh
-
-	// Stop all controllers.
-	var wg sync.WaitGroup
-	for _, pc := range mc.parentControllers {
-		wg.Add(1)
-		go func(pc *parentController) {
-			defer wg.Done()
-			pc.Stop()
-		}(pc)
-	}
-	wg.Wait()
-}
-
-func (mc *Metacontroller) processNextWorkItem() bool {
-	key, quit := mc.queue.Get()
-	if quit {
-		return false
-	}
-	defer mc.queue.Done(key)
-
-	err := mc.sync(key.(string))
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("failed to sync CompositeController %q: %v", key, err))
-		mc.queue.AddRateLimited(key)
-		return true
-	}
-
-	mc.queue.Forget(key)
-	return true
-}
-
-func (mc *Metacontroller) sync(key string) error {
-	_, name, err := cache.SplitMetaNamespaceKey(key)
-	if err != nil {
-		return err
-	}
-
+func (mc *Metacontroller) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	name := request.Name
 	klog.V(4).InfoS("Sync CompositeController", "name", name)
 
-	cc, err := mc.ccLister.Get(name)
+	cc := &v1alpha1.CompositeController{}
+	err := mc.k8s.Get(ctx.Background(), request.NamespacedName, cc)
```

Comment:
Should we write in an idiomatic style. For example:
```go
cc = v1alpha1.CompositeController{}
err := mc.k8s.Get(ctx.Background(), request.NamespacedName, &cc)
```

Reply:
I just wonder if we need another client or we can reuse `dynClient` already present there- i.e. :
```go
	resourceClient, _ := mc.dynClient.Kind(cc.APIVersion, cc.Kind)
	_, err := resourceClient.Namespace(request.Namespace).Get(request.Name, metav1.GetOptions{})
```
I will try it out locally and let you know

Reply:
Was able to do it via 
```go
resourceClient, err := mc.dynClient.Kind("metacontroller.k8s.io/v1alpha1", "CompositeController")
```
but it returns `Unstructured`. I will experiment a bit as  it it not very suitable I guess

Reply:
I think another difference is that the dynamic client is queuing the apiserver directly, while the one from controller-runtime is querying a cache. The cache is updates automatically in the background for each object that the controller is watching.